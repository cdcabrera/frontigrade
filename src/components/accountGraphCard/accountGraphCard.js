import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardGrid, Label as PFLabel, SparklineChart, Spinner } from 'patternfly-react';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';
import graphHelpers from '../../common/graphHelpers';
import accountGraphCardTypes from './accountGraphCardConstants';
import DropdownSelect from '../dropdownSelect/dropdownSelect';

class AccountGraphCard extends React.Component {
  componentDidMount() {
    const { filter, filterId, getAccountInstances, rhelOptions, rhocpOptions } = this.props;

    if (filter.query) {
      getAccountInstances(filterId, filter.query);
    }

    if (!filter.graphOpenshiftValue) {
      this.onSelectRhocpGraphFilter(rhocpOptions.find(option => option.active === true));
    }

    if (!filter.graphRhelValue) {
      this.onSelectRhelGraphFilter(rhelOptions.find(option => option.active === true));
    }
  }

  componentDidUpdate(prevProps) {
    const { filter, filterId, getAccountInstances, updateInstances } = this.props;

    if (updateInstances === true && updateInstances !== prevProps.updateInstances) {
      getAccountInstances(filterId, filter.query);
    }
  }

  onSelectRhelGraphFilter = option => {
    const { view, viewGlobal } = this.props;

    if (option) {
      store.dispatch({
        type: reduxTypes.filter.GRAPH_RHEL_SET_FILTER_VALUE,
        graphRhelValue: option.value,
        view,
        viewGlobal
      });
    }
  };

  onSelectRhocpGraphFilter = option => {
    const { view, viewGlobal } = this.props;

    if (option) {
      store.dispatch({
        type: reduxTypes.filter.GRAPH_OPENSHIFT_SET_FILTER_VALUE,
        graphOpenshiftValue: option.value,
        view,
        viewGlobal
      });
    }
  };

  renderPendingCard() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Card matchHeight accented className="cloudmeter-utilization-graph cloudmeter-utilization-graph-loading fadein">
          <Card.Body>
            <Spinner loading size="sm" className="blank-slate-pf-icon" />
            <div className="text-center">
              <small>Loading...</small>
            </div>
          </Card.Body>
        </Card>
      );
    }

    return null;
  }

  render() {
    const { error, filter, fulfilled, graphData, rhelOptions, rhocpOptions } = this.props;

    let chartTotals = {};
    let chartData = null;
    let displayRHELChartTotal = {};
    let displayRHELChartData = null;
    let displayRHOCPChartTotal = {};
    let displayRHOCPChartData = null;

    if (error) {
      return null;
    }

    if (graphData.dailyUsage && graphData.dailyUsage.length) {
      chartTotals = { ...graphHelpers.calculateGraphTotals({ ...graphData }) };
      chartData = graphHelpers.convertGraphData({ ...graphData });

      if (filter.graphOpenshiftValue) {
        displayRHOCPChartTotal = chartTotals[filter.graphOpenshiftValue];
        displayRHOCPChartData = chartData[filter.graphOpenshiftValue];
      }

      if (filter.graphRhelValue) {
        displayRHELChartTotal = chartTotals[filter.graphRhelValue];
        displayRHELChartData = chartData[filter.graphRhelValue];
      }
    }

    return (
      <CardGrid fluid matchHeight>
        <CardGrid.Row className="row-cards-pf">
          <CardGrid.Col sm={6}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight accented className="cloudmeter-utilization-graph fadein">
                <Card.Heading>
                  <DropdownSelect
                    id="graphRhel"
                    className="card-pf-time-frame-filter"
                    pullRight
                    onSelect={this.onSelectRhelGraphFilter}
                    options={rhelOptions}
                    selectValue={filter.graphRhelValue}
                  />
                  <Card.Title>Red Hat Enterprise Linux</Card.Title>
                </Card.Heading>
                <Card.Body>
                  <div className="cloudmeter-card-info">
                    <strong>{displayRHELChartTotal || 0}</strong>
                    <PFLabel bsStyle="warning">
                      <abbr title="Red Hat Enterprise Linux">RHEL</abbr>
                    </PFLabel>
                  </div>
                  {displayRHELChartData && (
                    <SparklineChart
                      key={filter.graphRhelValue}
                      data={displayRHELChartData}
                      tooltip={chartData.tooltips}
                    />
                  )}
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
          <CardGrid.Col sm={6}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight accented className="cloudmeter-utilization-graph fadein">
                <Card.Heading>
                  <DropdownSelect
                    id="graphOpenshift"
                    className="card-pf-time-frame-filter"
                    pullRight
                    onSelect={this.onSelectRhocpGraphFilter}
                    options={rhocpOptions}
                    selectValue={filter.graphOpenshiftValue}
                  />
                  <Card.Title>Red Hat OpenShift Container Platform</Card.Title>
                </Card.Heading>
                <Card.Body>
                  <div className="cloudmeter-card-info">
                    <strong>{displayRHOCPChartTotal || 0}</strong>
                    <PFLabel bsStyle="primary">
                      <abbr title="Red Hat OpenShift Container Platform">RHOCP</abbr>
                    </PFLabel>
                  </div>
                  {displayRHOCPChartData && (
                    <SparklineChart
                      key={filter.graphOpenshiftValue}
                      data={displayRHOCPChartData}
                      tooltip={chartData.tooltips}
                    />
                  )}
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
        </CardGrid.Row>
      </CardGrid>
    );
  }
}

AccountGraphCard.propTypes = {
  error: PropTypes.bool,
  filter: PropTypes.shape({
    graphOpenshiftValue: PropTypes.string,
    graphRhelValue: PropTypes.string,
    query: PropTypes.object.isRequired
  }).isRequired,
  filterId: PropTypes.number,
  fulfilled: PropTypes.bool,
  getAccountInstances: PropTypes.func,
  graphData: PropTypes.shape({
    dailyUsage: PropTypes.array,
    instancesOpenshift: PropTypes.number,
    instancesRhel: PropTypes.number
  }),
  pending: PropTypes.bool,
  rhelOptions: PropTypes.array,
  rhocpOptions: PropTypes.array,
  updateInstances: PropTypes.bool,
  view: PropTypes.string,
  viewGlobal: PropTypes.string
};

AccountGraphCard.defaultProps = {
  error: false,
  filterId: null,
  fulfilled: false,
  getAccountInstances: helpers.noop,
  graphData: {
    dailyUsage: [],
    instancesOpenshift: 0,
    instancesRhel: 0
  },
  pending: false,
  rhelOptions: accountGraphCardTypes.rhelOptions,
  rhocpOptions: accountGraphCardTypes.rhocpOptions,
  updateInstances: false,
  view: null,
  viewGlobal: null
};

const mapDispatchToProps = dispatch => ({
  getAccountInstances: (id, query) => dispatch(reduxActions.account.getAccountInstances(id, query))
});

const mapStateToProps = state => ({
  ...state.accountGraph
});

const ConnectedAccountGraphCard = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountGraphCard);

export { ConnectedAccountGraphCard as default, ConnectedAccountGraphCard, AccountGraphCard };
