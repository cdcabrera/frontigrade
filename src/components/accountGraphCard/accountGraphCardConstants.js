const rhelOptionsType = [
  { title: 'Core Hours', value: 'rhelVcpuTime' },
  { title: 'GB Memory Hours', value: 'rhelMemoryTime' },
  { title: 'Instance Hours', value: 'rhelRuntimeTime', active: true }
];

const rhocpOptionsType = [
  { title: 'Core Hours', value: 'openshiftVcpuTime', active: true },
  { title: 'GB Memory Hours', value: 'openshiftMemoryTime' },
  { title: 'Instance Hours', value: 'openshiftRuntimeTime' }
];

const accountGraphCardTypes = {
  rhelOptions: rhelOptionsType,
  rhocpOptions: rhocpOptionsType
};

export { accountGraphCardTypes as default, accountGraphCardTypes, rhelOptionsType, rhocpOptionsType };
