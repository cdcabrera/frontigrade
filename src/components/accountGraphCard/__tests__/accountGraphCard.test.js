import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedAccountGraphCard, AccountGraphCard } from '../accountGraphCard';

describe('AccountGraphCard Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component with default props', () => {
    const store = generateEmptyStore({ accountGraph: {} });
    const props = {
      filter: { query: {} },
      filterId: 1,
      view: 'test'
    };
    const component = shallow(<ConnectedAccountGraphCard {...props} />, { context: { store } });

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {
      filter: { query: {} },
      filterId: null
    };

    const component = mount(<AccountGraphCard {...props} />);

    expect(component.render()).toMatchSnapshot('non-connected');
  });

  it('should render multiple states', () => {
    const props = {
      filter: { query: {} },
      filterId: 1
    };

    const component = shallow(<AccountGraphCard {...props} />);

    component.setProps({
      error: true
    });

    expect(component).toMatchSnapshot('error');

    component.setProps({
      error: false,
      pending: true
    });

    expect(component).toMatchSnapshot('pending');

    component.setProps({
      error: false,
      pending: false,
      fulfilled: true
    });

    expect(component).toMatchSnapshot('fulfilled');
  });
});
