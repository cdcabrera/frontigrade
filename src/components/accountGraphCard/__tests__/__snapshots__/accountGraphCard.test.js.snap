// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountGraphCard Component should render a connected component with default props: connected 1`] = `
<AccountGraphCard
  error={false}
  filter={
    Object {
      "query": Object {},
    }
  }
  filterId={1}
  fulfilled={false}
  getAccountInstances={[Function]}
  graphData={
    Object {
      "dailyUsage": Array [],
      "instancesOpenshift": 0,
      "instancesRhel": 0,
    }
  }
  pending={false}
  rhelOptions={
    Array [
      Object {
        "title": "Core Hours",
        "value": "rhelVcpuTime",
      },
      Object {
        "title": "GB Memory Hours",
        "value": "rhelMemoryTime",
      },
      Object {
        "active": true,
        "title": "Instance Hours",
        "value": "rhelRuntimeTime",
      },
    ]
  }
  rhocpOptions={
    Array [
      Object {
        "active": true,
        "title": "Core Hours",
        "value": "openshiftVcpuTime",
      },
      Object {
        "title": "GB Memory Hours",
        "value": "openshiftMemoryTime",
      },
      Object {
        "title": "Instance Hours",
        "value": "openshiftRuntimeTime",
      },
    ]
  }
  updateInstances={false}
  view="test"
  viewGlobal={null}
/>
`;

exports[`AccountGraphCard Component should render a non-connected component: non-connected 1`] = `
<div
  class="container-cards-pf container-fluid"
>
  <div
    class=""
  >
    <div
      class="row-cards-pf row"
    >
      <div
        class="col-sm-6"
      />
      <div
        class="col-sm-6"
      />
    </div>
  </div>
</div>
`;

exports[`AccountGraphCard Component should render multiple states: error 1`] = `""`;

exports[`AccountGraphCard Component should render multiple states: fulfilled 1`] = `
<CardGrid
  className=""
  fluid={true}
  matchHeight={true}
>
  <Row
    bsClass="row"
    className="row-cards-pf"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={6}
    >
      <Card
        accented={true}
        aggregated={false}
        aggregatedMini={false}
        cardRef={null}
        className="cloudmeter-utilization-graph fadein"
        matchHeight={true}
      >
        <CardHeading
          className=""
        >
          <DropdownSelect
            className="card-pf-time-frame-filter"
            id="graphRhel"
            multiselect={false}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "title": "Core Hours",
                  "value": "rhelVcpuTime",
                },
                Object {
                  "title": "GB Memory Hours",
                  "value": "rhelMemoryTime",
                },
                Object {
                  "active": true,
                  "title": "Instance Hours",
                  "value": "rhelRuntimeTime",
                },
              ]
            }
            pullRight={true}
            selectValue={null}
            title="Select option"
          />
          <CardTitle
            className=""
          >
            Red Hat Enterprise Linux
          </CardTitle>
        </CardHeading>
        <CardBody
          className=""
        >
          <div
            className="cloudmeter-card-info"
          >
            <strong>
              <Component />
            </strong>
            <Label
              bsStyle="warning"
              type="default"
            >
              <abbr
                title="Red Hat Enterprise Linux"
              >
                RHEL
              </abbr>
            </Label>
          </div>
        </CardBody>
      </Card>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      sm={6}
    >
      <Card
        accented={true}
        aggregated={false}
        aggregatedMini={false}
        cardRef={null}
        className="cloudmeter-utilization-graph fadein"
        matchHeight={true}
      >
        <CardHeading
          className=""
        >
          <DropdownSelect
            className="card-pf-time-frame-filter"
            id="graphOpenshift"
            multiselect={false}
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "active": true,
                  "title": "Core Hours",
                  "value": "openshiftVcpuTime",
                },
                Object {
                  "title": "GB Memory Hours",
                  "value": "openshiftMemoryTime",
                },
                Object {
                  "title": "Instance Hours",
                  "value": "openshiftRuntimeTime",
                },
              ]
            }
            pullRight={true}
            selectValue={null}
            title="Select option"
          />
          <CardTitle
            className=""
          >
            Red Hat OpenShift Container Platform
          </CardTitle>
        </CardHeading>
        <CardBody
          className=""
        >
          <div
            className="cloudmeter-card-info"
          >
            <strong>
              <Component />
            </strong>
            <Label
              bsStyle="primary"
              type="default"
            >
              <abbr
                title="Red Hat OpenShift Container Platform"
              >
                RHOCP
              </abbr>
            </Label>
          </div>
        </CardBody>
      </Card>
    </Col>
  </Row>
</CardGrid>
`;

exports[`AccountGraphCard Component should render multiple states: pending 1`] = `
<CardGrid
  className=""
  fluid={true}
  matchHeight={true}
>
  <Row
    bsClass="row"
    className="row-cards-pf"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      sm={6}
    >
      <Card
        accented={true}
        aggregated={false}
        aggregatedMini={false}
        cardRef={null}
        className="cloudmeter-utilization-graph cloudmeter-utilization-graph-loading fadein"
        matchHeight={true}
      >
        <CardBody
          className=""
        >
          <Spinner
            className="blank-slate-pf-icon"
            inline={false}
            inverse={false}
            loading={true}
            size="sm"
          />
          <div
            className="text-center"
          >
            <small>
              Loading...
            </small>
          </div>
        </CardBody>
      </Card>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      sm={6}
    >
      <Card
        accented={true}
        aggregated={false}
        aggregatedMini={false}
        cardRef={null}
        className="cloudmeter-utilization-graph cloudmeter-utilization-graph-loading fadein"
        matchHeight={true}
      >
        <CardBody
          className=""
        >
          <Spinner
            className="blank-slate-pf-icon"
            inline={false}
            inverse={false}
            loading={true}
            size="sm"
          />
          <div
            className="text-center"
          >
            <small>
              Loading...
            </small>
          </div>
        </CardBody>
      </Card>
    </Col>
  </Row>
</CardGrid>
`;
